src/frontend/src/optimizer/rule/stream/filter_with_now_to_join_rule.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright 2023 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_common::types::DataType;
+use risingwave_pb::expr::expr_node::Type;
+use risingwave_pb::plan_common::JoinType;
+
+use crate::expr::{ExprRewriter, FunctionCall, InputRef};
+use crate::optimizer::plan_node::generic::GenericPlanRef;
+use crate::optimizer::plan_node::{LogicalFilter, LogicalJoin, LogicalNow};
+use crate::optimizer::rule::{BoxedRule, Rule};
+use crate::optimizer::PlanRef;
+use crate::utils::Condition;
+
+/// Convert `LogicalFilter` with now in predicate to left-semi `LogicalJoin`
+pub struct FilterWithNowToJoinRule {}
+impl Rule for FilterWithNowToJoinRule {
+    fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
+        let filter: &LogicalFilter = plan.as_logical_filter()?;
+        // if filter.predicate().conjunctions.iter().(|)
+
+        let lhs_len = filter.base.schema().len();
+
+        let mut now_filters = vec![];
+        let mut remainder = vec![];
+
+        // If the `now` is not a valid dynamic filter expression,
+        filter.predicate().conjunctions.iter().for_each(|expr| {
+            let mut rewriter = NowAsInputRef::new(lhs_len);
+            let expr = rewriter.rewrite_expr(expr.clone());
+            if rewriter.rewritten {
+                now_filters.push(expr);
+            } else {
+                remainder.push(expr);
+            }
+        });
+
+        if now_filters.is_empty() {
+            return None;
+        }
+        let mut new_plan = plan.inputs()[0].clone();
+
+        for now_filter in now_filters {
```

Comment:
I understand your idea. However, I still think that `now()-3<=t.a` should come first in `now()-3<=t.a AND t.a<=now()`

Reply:
`now()-3<=t.a` should also come first in `t.a<=now() AND now()-3<=t.a`

Reply:
Hmm, I agree. Since this PR is quite long, I can put it in a follow up PR and we should verify order in planner test.